lib/api/apiUtils/rateLimit/tokenBucket.js

```diff
@@ -0,0 +1,277 @@
+     */
+    async refill() {
+        this.refillInProgress = true;
+        const startTime = Date.now();
+
+        try {
+            // Calculate how many tokens to request
+            const requested = this.bufferSize - this.tokens;
+
+            if (requested <= 0) {
+                return; // Buffer already full
+            }
+
+            // Calculate GCRA parameters
+            const nodes = config.rateLimiting.nodes || 1;
+            const workers = config.clusters || 1;
+            const interval = calculateInterval(this.limitConfig.limit, nodes, workers);
+            const burstCapacity = (config.rateLimiting.bucket?.defaultConfig?.requestsPerSecond?.burstCapacity || 1) * 1000;
+
+            // Request tokens from Redis (atomic GCRA enforcement)
+            const granted = await new Promise((resolve, reject) => {
+                redisClient.grantTokens(
+                    this.bucketName,
+                    requested,
+                    interval,
+                    burstCapacity,
+                    (err, result) => {
+                        if (err) return reject(err);
+                        resolve(result);
+                    }
+                );
+            });
+
+            // Add granted tokens to buffer
+            this.tokens += granted;
+            this.refillCount++;
+
+            const duration = Date.now() - startTime;
+            this.lastRefillTime = Date.now();
+
+            this.log.debug('Token refill completed', {
+                bucketName: this.bucketName,
+                requested,
+                granted,
+                newBalance: this.tokens,
+                durationMs: duration,
+                refillCount: this.refillCount,
+            });
+
+            // Warn if refill took too long or granted too few
+            if (duration > 100) {
+                this.log.warn('Slow token refill detected', {
+                    bucketName: this.bucketName,
+                    durationMs: duration,
+                });
+            }
+
+            if (granted === 0 && requested > 0) {
+                this.log.trace('Token refill denied - quota exhausted', {
+                    bucketName: this.bucketName,
+                    requested,
+                });
+            }
+
+        } catch (err) {
+            this.log.error('Token refill failed', {
+                bucketName: this.bucketName,
+                error: err.message,
+                stack: err.stack,
+            });
+        } finally {
+            this.refillInProgress = false;
+        }
+    }
+
+    /**
+     * Record request timestamp for rate calculation
+     */
+    recordRequest() {
+        const now = Date.now();
+        this.requestTimestamps.push(now);
+
+        // Keep only last 1 second of timestamps
+        const cutoff = now - 1000;
+        while (this.requestTimestamps.length > 0 && this.requestTimestamps[0] < cutoff) {
+            this.requestTimestamps.shift();
```

Comment:
This operation could be slow-ish, if you have e.g. 1000 requests timestamps already in the array (with a client doing 1000 req/s) - you could instead use a `Set`, which can be iterated in insertion order, so deletions from it would be done in reasonable time, unnoticeable in practice.